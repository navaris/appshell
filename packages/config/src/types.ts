import { JSONSchema4, JSONSchema6, JSONSchema7 } from 'json-schema';

export type Schema = JSONSchema4 | JSONSchema6 | JSONSchema7;

export type ConfigValidator = {
  validate: <T>(...config: T[]) => void;
};

export type ConfigMap = Record<string, string>;

/* appshell.config.yaml types */
export type AppshellConfigRemote<TMetadata = Record<string, unknown>> = {
  id: string;
  url: string;
  metadata: TMetadata;
};

export type AppshellConfig<TMetadata = Record<string, unknown>> = {
  name?: string;
  remotes?: Record<string, AppshellConfigRemote<TMetadata>>;
  module: ModuleFederationPluginOptions;
};

/** Appshell manifest types */
export type AppshellRemote<TMetadata = Record<string, unknown>> = {
  id: string;
  url: string;
  scope: string;
  module: string;
  shareScope?: string;
  metadata: TMetadata;
};

export type AppshellManifest<TMetadata = Record<string, unknown>> = {
  remotes: Record<string, AppshellRemote<TMetadata>>;
  modules: Record<string, ModuleFederationPluginOptions>;
};

/**
 * Advanced configuration for modules that should be exposed by this container.
 */
export interface ExposesConfig {
  /**
   * Request to a module that should be exposed by this container.
   */
  import: string | string[];

  /**
   * Custom chunk name for the exposed module.
   */
  name?: string;
}

export interface ExposesObject {
  [index: string]: string | ExposesConfig | string[];
}

/**
 * Set explicit comments for `commonjs`, `commonjs2`, `amd`, and `root`.
 */
export interface LibraryCustomUmdCommentObject {
  /**
   * Set comment for `amd` section in UMD.
   */
  amd?: string;

  /**
   * Set comment for `commonjs` (exports) section in UMD.
   */
  commonjs?: string;

  /**
   * Set comment for `commonjs2` (module.exports) section in UMD.
   */
  commonjs2?: string;

  /**
   * Set comment for `root` (global variable) section in UMD.
   */
  root?: string;
}

/**
 * Description object for all UMD variants of the library name.
 */
export interface LibraryCustomUmdObject {
  /**
   * Name of the exposed AMD library in the UMD.
   */
  amd?: string;

  /**
   * Name of the exposed commonjs export in the UMD.
   */
  commonjs?: string;

  /**
   * Name of the property exposed globally by a UMD library.
   */
  root?: string | string[];
}

/**
 * Options for library.
 */
export interface LibraryOptions {
  /**
   * Add a comment in the UMD wrapper.
   */
  auxiliaryComment?: string | LibraryCustomUmdCommentObject;

  /**
   * Specify which export should be exposed as library.
   */
  export?: string | string[];

  /**
   * The name of the library (some types allow unnamed libraries too).
   */
  name?: string | string[] | LibraryCustomUmdObject;

  /**
   * Type of library (types included by default are 'var', 'module', 'assign', 'assign-properties', 'this', 'window', 'self', 'global', 'commonjs', 'commonjs2', 'commonjs-module', 'commonjs-static', 'amd', 'amd-require', 'umd', 'umd2', 'jsonp', 'system', but others might be added by plugins).
   */
  type: string;

  /**
   * If `output.libraryTarget` is set to umd and `output.library` is set, setting this to true will name the AMD module.
   */
  umdNamedDefine?: boolean;
}

export interface SharedConfig {
  /**
   * Include the provided and fallback module directly instead behind an async request. This allows to use this shared module in initial load too. All possible shared modules need to be eager too.
   */
  eager?: boolean;

  /**
   * Provided module that should be provided to share scope. Also acts as fallback module if no shared module is found in share scope or version isn't valid. Defaults to the property name.
   */
  import?: string | false;

  /**
   * Package name to determine required version from description file. This is only needed when package name can't be automatically determined from request.
   */
  packageName?: string;

  /**
   * Version requirement from module in share scope.
   */
  requiredVersion?: string | false;

  /**
   * Module is looked up under this key from the share scope.
   */
  shareKey?: string;

  /**
   * Share scope name.
   */
  shareScope?: string;

  /**
   * Allow only a single version of the shared module in share scope (disabled by default).
   */
  singleton?: boolean;

  /**
   * Do not accept shared module if version is not valid (defaults to yes, if local fallback module is available and shared module is not a singleton, otherwise no, has no effect if there is no required version specified).
   */
  strictVersion?: boolean;

  /**
   * Version of the provided module. Will replace lower matching versions, but not higher.
   */
  version?: string | false;
}

export interface SharedObject {
  [index: string]: string | SharedConfig;
}

/**
 * Advanced configuration for container locations from which modules should be resolved and loaded at runtime.
 */
export interface RemotesConfig {
  /**
   * Container locations from which modules should be resolved and loaded at runtime.
   */
  external: string | string[];

  /**
   * The name of the share scope shared with this remote.
   */
  shareScope?: string;
}

/**
 * Container locations from which modules should be resolved and loaded at runtime. Property names are used as request scopes.
 */
export interface RemotesObject {
  [index: string]: string | RemotesConfig | string[];
}

export interface ModuleFederationPluginOptions {
  /**
   * Modules that should be exposed by this container. When provided, property name is used as public name, otherwise public name is automatically inferred from request.
   */
  exposes?: (string | ExposesObject)[] | ExposesObject;

  /**
   * The filename of the container as relative path inside the `output.path` directory.
   */
  filename?: string;

  /**
   * Options for library.
   */
  library?: LibraryOptions;

  /**
   * The name of the container.
   */
  name?: string;

  /**
   * The external type of the remote containers.
   */
  remoteType?:
    | 'import'
    | 'var'
    | 'module'
    | 'assign'
    | 'this'
    | 'window'
    | 'self'
    | 'global'
    | 'commonjs'
    | 'commonjs2'
    | 'commonjs-module'
    | 'commonjs-static'
    | 'amd'
    | 'amd-require'
    | 'umd'
    | 'umd2'
    | 'jsonp'
    | 'system'
    | 'promise'
    | 'script'
    | 'node-commonjs';

  /**
   * Container locations and request scopes from which modules should be resolved and loaded at runtime. When provided, property name is used as request scope, otherwise request scope is automatically inferred from container location.
   */
  remotes?: (string | RemotesObject)[] | RemotesObject;

  /**
   * The name of the runtime chunk. If set a runtime chunk with this name is created or an existing entrypoint is used as runtime.
   */
  runtime?: string | false;

  /**
   * Share scope name used for all shared modules (defaults to 'default').
   */
  shareScope?: string;

  /**
   * Modules that should be shared in the share scope. When provided, property names are used to match requested modules in this compilation.
   */
  shared?: (string | SharedObject)[] | SharedObject;
}
