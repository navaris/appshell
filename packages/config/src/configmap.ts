import { entries, set, values } from 'lodash';
import { AppshellConfig, ConfigMap } from './types';

const matchVariable = /(\${\w+})/;
const matchWord = /[^\w @]/g;

const findVariables = (val: string) =>
  val
    .split(matchVariable)
    .filter((w) => w.match(matchVariable))
    ?.map((w) => w.replace(matchWord, ''));

const findVariablePlaceholders = (
  obj: object | string | number | undefined,
  results: Record<string, string> = {},
): Record<string, string> =>
  values(obj).reduce((acc: Record<string, string>, val: object | string | number | undefined) => {
    if (typeof val === 'string') {
      const VARS = findVariables(val);
      return VARS.length > 0 ? VARS.reduce((a, v) => set<string>(a, v, process.env[v]), acc) : acc;
    }
    if (typeof val === 'object') {
      return findVariablePlaceholders(val, acc);
    }
    return acc;
  }, results);

const apply = <T extends object>(obj: T, configMap: ConfigMap): T => {
  entries(obj).forEach(([key, val]) => {
    if (typeof val === 'string') {
      const VARS = findVariables(val);
      if (VARS.length > 0) {
        VARS.forEach((v) => {
          const cur = obj[key];
          const value = configMap[v];
          if (!value) {
            // eslint-disable-next-line no-console
            console.log(`warning: value for ${v} is ${value}`);
          }
          set<string>(obj, key, cur.replace(`$\{${v}}`, configMap[v]));
        });
      }
    } else if (typeof val === 'object') {
      apply(val, configMap);
    }
  });

  return obj;
};

const create = (config: AppshellConfig) => findVariablePlaceholders(config);

export default {
  apply,
  create,
};
