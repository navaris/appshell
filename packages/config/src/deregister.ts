/* eslint-disable no-param-reassign */
import fs from 'fs';
import axios from './axios';
import { AppshellManifest, AppshellRegister } from './types';
import { isValidUrl } from './utils';
import * as validators from './validators';

const removeFromRegister = (doc: AppshellRegister, key: string) => {
  if (doc.index[key]) {
    delete doc.index[key];
  } else {
    // eslint-disable-next-line no-console
    console.warn(`index entry not found for '${key}'`);
  }

  if (doc.metadata && doc.metadata[key]) {
    delete doc.metadata[key];
  } else {
    // eslint-disable-next-line no-console
    console.warn(`metadata entry not found for '${key}'`);
  }

  return doc;
};

const removeFromManifest = (doc: AppshellManifest, key: string) => {
  const targetModule = key.split('/')[0];

  if (!doc.remotes[key]) {
    // eslint-disable-next-line no-console
    console.warn(`manifest entry not found for remotes[${key}]`);
  }

  if (!doc.modules[targetModule]) {
    // eslint-disable-next-line no-console
    console.warn(`manifest entry not found for modules[${key}]`);
  }

  if (!doc.environment[targetModule]) {
    // eslint-disable-next-line no-console
    console.warn(`manifest entry not found for environment[${key}]`);
  }

  delete doc.remotes[key];
  delete doc.modules[targetModule];
  delete doc.environment[targetModule];

  return doc;
};

const updateRegister = (registry: string, key: string) => {
  const file = `${registry}/appshell.register.json`;

  if (!fs.existsSync(file)) {
    // eslint-disable-next-line no-console
    console.warn(`registry file not found ${file}`);
    return;
  }

  const register: AppshellRegister = JSON.parse(fs.readFileSync(file, 'utf-8'));
  const doc = removeFromRegister(register, key);

  validators.appshell_register.validate(doc);

  fs.writeFileSync(file, JSON.stringify(doc));
};

const updateManifest = (registry: string, key: string) => {
  const file = `${registry}/appshell.manifest.json`;

  if (!fs.existsSync(file)) {
    // eslint-disable-next-line no-console
    console.warn(`registry file not found ${file}`);
    return;
  }

  const manifest: AppshellManifest = JSON.parse(fs.readFileSync(file, 'utf-8'));
  const doc = removeFromManifest(manifest, key);

  validators.merge_manifests.validate(doc);

  fs.writeFileSync(file, JSON.stringify(doc));
};

export default async (manifestKey: string, registry: string) => {
  // eslint-disable-next-line no-console
  console.log(`deregistering manifest from registry ${registry}`);

  if (isValidUrl(registry)) {
    await axios.delete(registry, { data: { manifestKey } });

    return;
  }

  if (!fs.existsSync(registry)) {
    // eslint-disable-next-line no-console
    console.warn(`registry location not found ${registry}`);
  }

  updateRegister(registry, manifestKey);
  updateManifest(registry, manifestKey);
};
