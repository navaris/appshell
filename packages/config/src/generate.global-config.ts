/* eslint-disable no-console */
import { AppshellGlobalConfig } from './types';
import { isValidUrl, merge } from './utils';
import loadJson from './utils/loadJson';
import { AppshellGlobalConfigValidator } from './validators';

type GenerateGlobalConfigOptions = {
  insecure: boolean;
};

export default async (
  registries: string[],
  options: GenerateGlobalConfigOptions = { insecure: false },
): Promise<AppshellGlobalConfig> => {
  const defaultGlobalConfig = { index: {}, metadata: {}, overrides: { environment: {} } };

  if (registries.length < 1) {
    console.log(`No registries found. skipping global configuration generation.`);
    return defaultGlobalConfig;
  }

  console.log(
    `generating global appshell configuration --registry=${JSON.stringify(registries, null, 2)}`,
  );

  try {
    const registers = await Promise.all(
      registries.map((reg) => {
        const registry = isValidUrl(reg) ? `${reg}/appshell.config.json` : reg;
        return loadJson<AppshellGlobalConfig>(registry, {
          insecure: options.insecure,
          target: /(.config.json)/i,
        });
      }),
    ).then((items) => items.flat());
    console.log(
      `Generating global configuration from ${registers.length} source${
        registers.length === 1 ? '' : 's'
      }`,
    );

    return merge(AppshellGlobalConfigValidator, ...registers) as AppshellGlobalConfig;

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  } catch (err: any) {
    console.error('Error generating global appshell configuration', err.message);
  }

  return defaultGlobalConfig;
};
