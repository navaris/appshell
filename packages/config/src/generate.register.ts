/* eslint-disable no-console */
import { AppshellRegister } from './types';
import { isValidUrl, merge } from './utils';
import loadJson from './utils/loadJson';
import { appshell_register } from './validators';

type GenerateRegisterOptions = {
  insecure: boolean;
};

export default async (
  registries: string[],
  options: GenerateRegisterOptions = { insecure: false },
): Promise<AppshellRegister> => {
  const defaultRegister = { index: {}, metadata: {}, overrides: { environment: {} } };

  if (registries.length < 1) {
    console.log(`No registries found. skipping register generation.`);
    return defaultRegister;
  }

  console.log(`generating appshell register --registry=${JSON.stringify(registries, null, 2)}`);

  try {
    const registers = await Promise.all(
      registries.map((reg) => {
        const registry = isValidUrl(reg) ? `${reg}/appshell.register.json` : reg;
        return loadJson<AppshellRegister>(registry, {
          insecure: options.insecure,
          target: /(.register.json)/i,
        });
      }),
    ).then((items) => items.flat());
    console.log(
      `Generating register from ${registers.length} source${registers.length === 1 ? '' : 's'}`,
    );

    return merge(appshell_register, ...registers) as AppshellRegister;

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  } catch (err: any) {
    console.error('Error generating appshell index', err.message);
  }

  return defaultRegister;
};
