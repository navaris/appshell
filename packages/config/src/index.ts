/**
 * @appshell/config package API
 */
export { default as configmap } from './configmap';
export { default as deregister } from './deregister';
export { default as generateEnv } from './generate.env';
export { default as generateGlobalConfig } from './generate.global-config';
export { default as generateManifest } from './generate.manifest';
export { default as register } from './register';
export type {
  AppshellConfig,
  AppshellConfigRemote,
  AppshellIndex,
  AppshellManifest,
  AppshellRemote,
  AppshellTemplate,
  Metadata,
  Schema,
} from './types';
export * as utils from './utils';
export * as validators from './validators';
