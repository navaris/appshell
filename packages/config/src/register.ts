import fs from 'fs';
import axios from './axios';
import { AppshellGlobalConfig, AppshellManifest, ConfigValidator } from './types';
import { isValidUrl, merge } from './utils';
import * as validators from './validators';

const updateDocument = (file: string, data: object, validator: ConfigValidator) => {
  if (!fs.existsSync(file)) {
    fs.writeFileSync(file, JSON.stringify(data));
    return;
  }

  const doc = JSON.parse(fs.readFileSync(file, 'utf-8'));
  fs.writeFileSync(file, JSON.stringify(merge(validator, doc, data)));
};

export default async (manifest: AppshellManifest, registryPathOrUrl: string) => {
  // eslint-disable-next-line no-console
  console.log(`registering manifest to registry ${registryPathOrUrl}`);

  if (isValidUrl(registryPathOrUrl)) {
    await axios.post(registryPathOrUrl, manifest);

    return;
  }

  if (!fs.existsSync(registryPathOrUrl)) {
    fs.mkdirSync(registryPathOrUrl, { recursive: true });
  }

  const registry = Object.entries(manifest.remotes).reduce(
    (acc, [key, remote]) => {
      acc.index[key] = remote.manifestUrl;
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      acc.metadata![key] = remote.metadata;

      return acc;
    },
    {
      index: {},
      metadata: {},
      overrides: process.env.NODE_ENV !== 'production' ? manifest.overrides : {},
    } as AppshellGlobalConfig,
  );

  updateDocument(
    `${registryPathOrUrl}/appshell.config.json`,
    registry,
    validators.AppshellGlobalConfigValidator,
  );
  updateDocument(
    `${registryPathOrUrl}/appshell.snapshot.json`,
    manifest,
    validators.AppshellManifestValidator,
  );
};
