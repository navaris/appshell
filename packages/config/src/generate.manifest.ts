/* eslint-disable no-param-reassign */
import fs from 'fs';
import path from 'path';
import configmap from './configmap';
import { toAppshellManifest } from './mappers/appshell.config';
import { AppshellTemplate } from './types';

/**
 * Generates an appshell manifest
 * @param templatePath path of the appshell manifest template to compile into an appshell manifest
 * @returns an appshell manifest
 */
export default async <TMetadata extends Record<string, unknown>>(templatePath: string) => {
  if (!fs.existsSync(templatePath)) {
    // eslint-disable-next-line no-console
    console.log(`No template file found '${templatePath}'`);

    return null;
  }

  const file = fs.readFileSync(path.resolve(templatePath), 'utf-8');
  const template = JSON.parse(file) as AppshellTemplate;
  const configMap = configmap.create(template);

  return toAppshellManifest<TMetadata>(template, configMap);
};
