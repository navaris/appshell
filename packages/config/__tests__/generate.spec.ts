import { HttpStatusCode } from 'axios';
import https from 'https';
import { keys, values } from 'lodash';
import path from 'path';
import generateEnv from '../src/generate.env';
import generateGlobalConfig from '../src/generate.global-config';
import generate from '../src/generate.manifest';
import * as utils from '../src/utils';
import manifest from './assets/appshell.manifest.json';
import mockAxios from './utils/axios';

type TestMetadata = {
  route: string;
  displayName: string;
  displayGroup: string;
  order: number;
  icon: string;
};

describe('generate', () => {
  const packageName = 'config';

  describe('runtime env file', () => {
    const env = `packages/${packageName}/__tests__/assets/test.env`;

    it('should generate the runtime environment js file', async () => {
      const environment = await generateEnv(env);

      expect(Object.fromEntries(environment)).toStrictEqual({
        REGISTRY: 'packages/cli/__tests__/assets/appshell_registry',
        ROOT: 'TestModule/Workspace',
        TEST_ENV_FOO: 'foo',
        TEST_ENV_BAR: 'bar',
      });
    });

    it('should capture only prefixed environment vars when prefix is supplied', async () => {
      const environment = await generateEnv(env, 'TEST_');

      expect(Object.fromEntries(environment)).toStrictEqual({
        TEST_ENV_FOO: 'foo',
        TEST_ENV_BAR: 'bar',
      });
    });

    it('should fail gracefully when env file is not found', async () => {
      await expect(generateEnv('does_not_exist.env')).rejects.toThrow();
    });
  });

  describe('manifest', () => {
    const configTemplate = path.resolve(
      `packages/${packageName}/__tests__/assets/appshell.template.json`,
    );

    process.env.APPS_TEST_URL = 'http://remote-module.com/remoteEntry.js';
    process.env.RUNTIME_ENV = 'development';
    process.env.RUNTIME_ENV_VERSION = '1.0.0';

    it('should generate an appshell manifest from the config template', async () => {
      const config = await generate(configTemplate);

      expect(config).toMatchSnapshot();
    });

    it('should produce null when config is available to process', async () => {
      const config = await generate(undefined as any);

      expect(config).toBeNull();
    });

    it('should contain all remotes', async () => {
      const config = await generate(configTemplate);
      const expectedRemotes = keys(manifest.remotes);
      const actualRemotes = keys(config?.remotes);

      expect(expectedRemotes).toEqual(actualRemotes);
    });

    it('should apply environment variables to configuration', async () => {
      const config = await generate(configTemplate);

      const actualUrls = values(config?.remotes).flatMap((remote) => remote.manifestUrl);

      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      expect(actualUrls.some((url) => url.includes(process.env.APPS_TEST_URL!))).toBeTruthy();
    });

    it('should capture metadata', async () => {
      const config = await generate<TestMetadata>(configTemplate);

      expect(values(config?.remotes).flatMap((remote) => remote.metadata)).toHaveLength(3);
    });
  });

  describe('register', () => {
    let agentConstructorSpy: jest.SpyInstance;
    let consoleSpy: jest.SpyInstance;
    beforeEach(() => {
      agentConstructorSpy = jest.spyOn(https, 'Agent').mockReturnThis();
      consoleSpy = jest.spyOn(console, 'error');
    });

    afterEach(() => {
      agentConstructorSpy.mockRestore();
      consoleSpy.mockRestore();
    });

    const testGlobalConfig = [
      `packages/${packageName}/__tests__/assets/registries/appshell.config.json`,
    ];
    const adjunctRegistries = [
      `packages/${packageName}/__tests__/assets/registries/registry_a`,
      `packages/${packageName}/__tests__/assets/registries/registry_b`,
    ];
    const remoteRegistries = ['https://test.com/registry'];

    it('should merge multiple valid global configurations', async () => {
      const config = await generateGlobalConfig([...adjunctRegistries, ...testGlobalConfig]);

      expect(config).toMatchSnapshot();
    });

    it('should handle a single global configuration file', async () => {
      const config = await generateGlobalConfig(testGlobalConfig);

      expect(config).toMatchSnapshot();
    });

    it('should handle a directory', async () => {
      const config = await generateGlobalConfig(adjunctRegistries);

      expect(config).toMatchSnapshot();
    });

    it('should log error and return empty if something goes wrong', async () => {
      jest.spyOn(utils, 'merge').mockImplementationOnce(() => {
        throw new Error('Something went wrong');
      });
      const config = await generateGlobalConfig(testGlobalConfig);

      expect(consoleSpy).toHaveBeenCalledWith(
        'Error generating global appshell configuration',
        'Something went wrong',
      );
      expect(config).toMatchObject({});
    });

    it('should configure http client when insecure is true', async () => {
      mockAxios
        .onGet(/\/registry/i)
        .reply(HttpStatusCode.Ok, { status: HttpStatusCode.Ok, statusText: 'OK', data: {} });

      await generateGlobalConfig(remoteRegistries, { insecure: true });

      expect(agentConstructorSpy).toHaveBeenCalledWith({ rejectUnauthorized: false });
    });

    it('should not configure http client when insecure is false', async () => {
      mockAxios
        .onGet(/\/registry/i)
        .reply(HttpStatusCode.Ok, { status: HttpStatusCode.Ok, statusText: 'OK', data: {} });

      await generateGlobalConfig(remoteRegistries, { insecure: false });

      expect(agentConstructorSpy).not.toHaveBeenCalled();
    });
  });
});
