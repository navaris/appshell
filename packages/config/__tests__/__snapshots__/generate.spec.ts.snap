// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate manifest should generate an appshell manifest from the config template 1`] = `
{
  "environment": {
    "RUNTIME_ENV": "development",
    "RUNTIME_ENV_VERSION": "1.0.0",
  },
  "modules": {
    "TestModule": {
      "exposes": {
        "./Bar": "./src/Entry2",
        "./Foo": "./src/Entry1",
      },
      "filename": "remoteEntry.js",
      "library": {
        "name": "some shared lib",
        "type": "umd",
      },
      "name": "TestModule",
      "remoteType": "amd",
      "remotes": {
        "./Remote1": {
          "external": "external module",
          "shareScope": "@shared",
        },
      },
      "runtime": "runtimeChunk",
      "shareScope": "@sharedscope",
      "shared": {
        "package1": {
          "requiredVersion": "0.1.0",
          "singleton": true,
        },
        "package2": {
          "requiredVersion": "1.1.0",
          "singleton": true,
        },
      },
    },
  },
  "remotes": {
    "BizModule/Biz": {
      "id": "BizModule/Biz",
      "manifestUrl": "http://localhost:4040/appshell.manifest.json",
      "metadata": {
        "displayGroup": "auxiliary",
        "displayName": "Biz App",
        "icon": "ViewList",
        "order": 30,
        "route": "/biz",
      },
      "module": "./Biz",
      "remoteEntryUrl": "http://localhost:4040/remoteEntry.js",
      "scope": "BizModule",
      "shareScope": "@sharedscope",
    },
    "TestModule/Bar": {
      "id": "TestModule/Bar",
      "manifestUrl": "http://remote-module.com/remoteEntry.js/appshell.manifest.json",
      "metadata": {
        "displayGroup": "main",
        "displayName": "Bar App",
        "icon": "ViewList",
        "order": 20,
        "route": "/bar",
      },
      "module": "./Bar",
      "remoteEntryUrl": "http://remote-module.com/remoteEntry.js/undefined",
      "scope": "TestModule",
      "shareScope": "@sharedscope",
    },
    "TestModule/Foo": {
      "id": "TestModule/Foo",
      "manifestUrl": "http://remote-module.com/remoteEntry.js/appshell.manifest.json",
      "metadata": {
        "displayGroup": "main",
        "displayName": "Foo App",
        "icon": "ViewList",
        "order": 10,
        "route": "/foo",
      },
      "module": "./Foo",
      "remoteEntryUrl": "http://remote-module.com/remoteEntry.js/remoteEntry.js",
      "scope": "TestModule",
      "shareScope": "@sharedscope",
    },
  },
}
`;

exports[`generate register should handle a directory 1`] = `
{
  "index": {
    "ContainerModule/Container": "http://localhost:30001/appshell.manifest.json",
    "PingModule/Ping": "http://localhost:30011/appshell.manifest.json",
    "PongModule/CoolComponent": "http://localhost:30021/appshell.manifest.json",
    "PongModule/Pong": "http://localhost:30021/appshell.manifest.json",
    "TestModule1/Component1": "http://test.com/1/appshell.manifest.json",
    "TestModule1/Component2": "http://test.com/1/appshell.manifest.json",
    "TestModule2/Component": "http://test.com/2/appshell.manifest.json",
    "TestModule3/ComponentA": "http://test.com/3/appshell.manifest.json",
    "TestModule3/ComponentB": "http://test.com/3/appshell.manifest.json",
  },
  "metadata": {
    "ContainerModule/Container": {
      "displayGroup": "main",
      "displayName": "Host App",
      "icon": "ViewList",
      "order": 0,
      "route": "/",
    },
    "PingModule/Ping": {
      "displayGroup": "main",
      "displayName": "Ping Micro-Frontend",
      "icon": "ViewList",
      "order": 10,
      "route": "/ping",
    },
    "PongModule/CoolComponent": {
      "displayGroup": "main",
      "displayName": "Cool Shared Component",
      "icon": "ViewList",
      "order": 30,
      "route": "/cool",
    },
    "PongModule/Pong": {
      "displayGroup": "main",
      "displayName": "Pong Micro-Frontend",
      "icon": "ViewList",
      "order": 20,
      "route": "/pong",
    },
    "TestModule1/Component1": {
      "foo": "bar",
    },
    "TestModule1/Component2": {
      "biz": "bang",
    },
    "TestModule2/Component": {
      "arg1": "val1",
      "arg2": "val2",
      "arg3": "val3",
    },
    "TestModule3/ComponentA": {
      "argA": "valA",
      "argB": "valB",
    },
    "TestModule3/ComponentB": {
      "argC": "valC",
      "argD": "valD",
    },
  },
}
`;

exports[`generate register should handle a single global configuration file 1`] = `
{
  "index": {
    "ContainerModule/Container": "http://localhost:30001/appshell.manifest.json",
    "PingModule/Ping": "http://localhost:30011/appshell.manifest.json",
    "PongModule/CoolComponent": "http://localhost:30021/appshell.manifest.json",
    "PongModule/Pong": "http://localhost:30021/appshell.manifest.json",
  },
  "metadata": {
    "ContainerModule/Container": {
      "displayGroup": "main",
      "displayName": "Host App",
      "icon": "ViewList",
      "order": 0,
      "route": "/",
    },
    "PingModule/Ping": {
      "displayGroup": "main",
      "displayName": "Ping Micro-Frontend",
      "icon": "ViewList",
      "order": 10,
      "route": "/ping",
    },
    "PongModule/CoolComponent": {
      "displayGroup": "main",
      "displayName": "Cool Shared Component",
      "icon": "ViewList",
      "order": 30,
      "route": "/cool",
    },
    "PongModule/Pong": {
      "displayGroup": "main",
      "displayName": "Pong Micro-Frontend",
      "icon": "ViewList",
      "order": 20,
      "route": "/pong",
    },
  },
  "overrides": {
    "environment": {
      "ContainerModule": {
        "ENV_VAR_A": "Overridden value of A",
      },
    },
  },
}
`;

exports[`generate register should merge multiple valid global configurations 1`] = `
{
  "index": {
    "ContainerModule/Container": "http://localhost:30001/appshell.manifest.json",
    "PingModule/Ping": "http://localhost:30011/appshell.manifest.json",
    "PongModule/CoolComponent": "http://localhost:30021/appshell.manifest.json",
    "PongModule/Pong": "http://localhost:30021/appshell.manifest.json",
    "TestModule1/Component1": "http://test.com/1/appshell.manifest.json",
    "TestModule1/Component2": "http://test.com/1/appshell.manifest.json",
    "TestModule2/Component": "http://test.com/2/appshell.manifest.json",
    "TestModule3/ComponentA": "http://test.com/3/appshell.manifest.json",
    "TestModule3/ComponentB": "http://test.com/3/appshell.manifest.json",
  },
  "metadata": {
    "ContainerModule/Container": {
      "displayGroup": "main",
      "displayName": "Host App",
      "icon": "ViewList",
      "order": 0,
      "route": "/",
    },
    "PingModule/Ping": {
      "displayGroup": "main",
      "displayName": "Ping Micro-Frontend",
      "icon": "ViewList",
      "order": 10,
      "route": "/ping",
    },
    "PongModule/CoolComponent": {
      "displayGroup": "main",
      "displayName": "Cool Shared Component",
      "icon": "ViewList",
      "order": 30,
      "route": "/cool",
    },
    "PongModule/Pong": {
      "displayGroup": "main",
      "displayName": "Pong Micro-Frontend",
      "icon": "ViewList",
      "order": 20,
      "route": "/pong",
    },
    "TestModule1/Component1": {
      "foo": "bar",
    },
    "TestModule1/Component2": {
      "biz": "bang",
    },
    "TestModule2/Component": {
      "arg1": "val1",
      "arg2": "val2",
      "arg3": "val3",
    },
    "TestModule3/ComponentA": {
      "argA": "valA",
      "argB": "valB",
    },
    "TestModule3/ComponentB": {
      "argC": "valC",
      "argD": "valD",
    },
  },
  "overrides": {
    "environment": {
      "ContainerModule": {
        "ENV_VAR_A": "Overridden value of A",
      },
    },
  },
}
`;
